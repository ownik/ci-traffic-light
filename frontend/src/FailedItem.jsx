import React from 'react';
import PropTypes from 'prop-types';
import ProgressSpinner from './ProgressSpinner';

const FailedItem = ({ displayName, href, running }) => (
  <li>
    <span>
      {!href || href === '' ? displayName : <a href={href}>{displayName}</a>}
    </span>
    {running && <ProgressSpinner />}
  </li>
);

FailedItem.propTypes = {
  displayName: PropTypes.string,
  href: PropTypes.string,
  running: PropTypes.bool,
};

FailedItem.defaultProps = {
  href: '',
};

export default FailedItem;
