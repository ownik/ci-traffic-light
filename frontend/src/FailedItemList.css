div.failed-item-list {
  visibility: visible;
  display: block;
  color: white;
  font-size: 4em;
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}

div.failed-item-list > ul {
  list-style: none;
}

div.failed-item-list > ul > li {
  transition: opacity 1s;
}

div.failed-item-list > ul > li.enter {
  opacity: 0;
}

div.failed-item-list > ul > li.enter-active {
  opacity: 1;
}

div.failed-item-list > ul > li.exit {
  opacity: 1;
}

div.failed-item-list > ul > li.exit-active {
  opacity: 0;
}

div.failed-item-list > ul > li > span > a:link,
a:visited {
  color: white;
  text-decoration: none;
}

div.failed-item-list > ul > li > span > a:hover,
a:active {
  color: white;
  text-decoration: none;
  border-bottom: 2px solid white;
  display: inline-block;
  line-height: 0.85;
}
