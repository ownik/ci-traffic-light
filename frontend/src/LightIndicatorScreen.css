div.indicator-screen {
  height: 100vh;
  margin: 0;
  transition: all 1s ease;
}

div.indicator-screen.success {
  background: linear-gradient(#11998e, #38ef7d);
}
div.indicator-screen.investigation {
  background: linear-gradient(#ffc11e, #ffd600);
}
div.indicator-screen.fail {
  background: linear-gradient(#ff416c, #ff4b2b);
}
div.indicator-screen.idle {
  background: linear-gradient(-45deg, #ee7752, #e73c7e, #23a6d5, #23d5ab);
  background-size: 400% 400%;
  animation: gradient 15s ease infinite;
}

@keyframes gradient {
  0% {
    background-position: 0% 50%;
  }
  50% {
    background-position: 100% 50%;
  }
  100% {
    background-position: 0% 50%;
  }
}
