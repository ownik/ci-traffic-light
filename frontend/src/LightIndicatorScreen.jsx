import React from 'react';
import PropTypes from 'prop-types';
import FailedItemList from './FailedItemList';

import './LightIndicatorScreen.css';

const LightIncicatorScreen = ({ status, items, children }) => (
  <div className={`indicator-screen ${status}`}>
    {children}
    <FailedItemList items={items} />
  </div>
);

LightIncicatorScreen.propTypes = {
  status: PropTypes.string,
  items: PropTypes.array,
  children: PropTypes.node,
};

LightIncicatorScreen.defaultProps = {
  status: 'idle',
  items: [],
  children: null,
};

export default LightIncicatorScreen;
