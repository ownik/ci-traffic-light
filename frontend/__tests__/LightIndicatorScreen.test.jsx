import React from 'react';
import { shallow } from 'enzyme';
import LightIndicatorScreen from '../src/LightIndicatorScreen';
import FailedItemList from '../src/FailedItemList';
import FailedItem from '../src/FailedItem';

describe('LightIndicatorScreen', () => {
  test('renders without crashing', () => {
    shallow(<LightIndicatorScreen />);
  });

  test('first element must be div', () => {
    const wrapper = shallow(<LightIndicatorScreen />);
    expect(wrapper.type()).toBe('div');
  });

  test('check colors props - default status', () => {
    const wrapper = shallow(<LightIndicatorScreen />);
    expect(wrapper.get(0).props.style).toEqual({
      background: 'linear-gradient(#FF416C, #FF4B2B)',
    });
  });

  test('check colors props - fail status', () => {
    const wrapper = shallow(<LightIndicatorScreen status="fail" />);
    expect(wrapper.get(0).props.style).toEqual({
      background: 'linear-gradient(#FF416C, #FF4B2B)',
    });
  });

  test('check colors props - success status', () => {
    const wrapper = shallow(<LightIndicatorScreen status="success" />);
    expect(wrapper.get(0).props.style).toEqual({
      background: 'linear-gradient(#11998E, #38EF7D)',
    });
  });

  test('correct render children', () => {
    const MockContainer = () => <div></div>;
    const wrapper = shallow(
      <LightIndicatorScreen>
        <MockContainer />
      </LightIndicatorScreen>
    );
    expect(wrapper.find(MockContainer)).toHaveLength(1);
  });

  test('className for status fail', () => {
    const wrapper = shallow(<LightIndicatorScreen status="fail" />);
    expect(wrapper.props()).toHaveProperty(
      'className',
      'indicator-screen fail'
    );
  });

  test('className for status success', () => {
    const wrapper = shallow(<LightIndicatorScreen status="success" />);
    expect(wrapper.props()).toHaveProperty(
      'className',
      'indicator-screen success'
    );
  });

  test('FailedItemList visible on status fail', () => {
    const wrapper = shallow(<LightIndicatorScreen status="fail" />);
    expect(wrapper.find(FailedItemList)).toHaveLength(1);
  });

  test('FailedItemList visible on status success', () => {
    const wrapper = shallow(<LightIndicatorScreen status="success" />);
    expect(wrapper.find(FailedItemList)).toHaveLength(1);
  });

  test('FailedItemList visible on status fail with setted items', () => {
    const items = [
      { displayName: 'item1' },
      { displayName: 'item2' },
      { displayName: 'item3' },
    ];
    const wrapper = shallow(
      <LightIndicatorScreen status="fail" items={items} />
    );
    const listWrapper = wrapper.find(FailedItemList).dive();
    expect(listWrapper.find(FailedItem)).toHaveLength(items.length);
    listWrapper.find(FailedItem).forEach((item, index) => {
      expect(item.dive().text()).toEqual(items[index].displayName);
    });
  });
});
