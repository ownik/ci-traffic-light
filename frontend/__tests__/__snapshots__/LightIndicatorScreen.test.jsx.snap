// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LightIndicatorScreen fail state snapshot 1`] = `"<div class=\\"indicator-screen fail\\" style=\\"background:linear-gradient(#FF416C, #FF4B2B)\\"><div class=\\"failed-item-list\\"><ul><li><span><a href=\\"testhref1\\">item1</a></span></li><li><span><a href=\\"testhref2\\">item2</a></span></li><li><span>item3</span></li></ul></div></div>"`;

exports[`LightIndicatorScreen success state snapshot 1`] = `"<div class=\\"indicator-screen success\\" style=\\"background:linear-gradient(#11998E, #38EF7D)\\"><div class=\\"failed-item-list\\"><ul></ul></div></div>"`;
